[Merged by Bors] - feat(algebra/monoid_algebra): adjointness for the functor `G ↦ monoid_algebra k G` when `G` carries only `has_mul`

ocfnash opened the pull request:
---
- [x] depends on: #7863

This replaces #6571

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

ocfnash committed:
feat(algebra/non_unital_alg_hom): define `non_unital_alg_hom`

The motivation is to be able to state the universal property
for a magma algebra using bundled morphisms.

ocfnash committed:
Add `to_non_unital_alg_hom`

ocfnash committed:
Generalise from `module` to just `distrib_mul_action`

ocfnash committed:
Golf and move `monoid_hom.inverse`

ocfnash committed:
Drop addition of `non_unital_alg_hom.lean`

The plan is to split this PR in two since the other changes are
deep and cause long build times.

ocfnash committed:
Revert "Drop addition of `non_unital_alg_hom.lean`"

This reverts commit 880e44bc0b8f11151f7df2f40b3540633c407a40.

ocfnash committed:
Merge branch 'master' into non_unital_alg_hom2

ocfnash committed:
Clean up after merge

ocfnash committed:
Fix lemma name

I missed this when I generalised from extending linear_map to extending distrib_mul_action_hom

ocfnash committed:
Merge branch 'master' into non_unital_alg_hom2

ocfnash committed:
feat(algebra/monoid_algebra): adjointness for the functor `G ↦ monoid_algebra k G` when `G` carries only `has_mul`

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -82,13 +86,88 @@ instance : non_unital_non_assoc_semiring (monoid_algebra k G) :=
   .. finsupp.add_comm_monoid }
 
+instance is_scalar_tower_self : is_scalar_tower k (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b, show (t • a) * b = t • (a * b),
```
Are you sure this isn't a less general case of `is_scalar_tower` below?

eric-wieser replied:
Ah, I'm thinking of a generalized multiplication we don't have yet

ocfnash replied:
I believe this is a different instance because it makes a statement about the convolution multiplication `monoid_algebra.has_mul` whereas the instance [here](https://github.com/leanprover-community/mathlib/blob/461b444d065e1747f1bda60070b4b5d6a3026fb2/src/algebra/monoid_algebra.lean#L240) is a statement about the pointwise multiplication defined in `finsupp`.

Note also that to apply the instance below, we would need a `monoid` structure on `monoid_algebra k G` which we do not have.

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -62,6 +63,9 @@ namespace monoid_algebra
 variables {k G}
 
+instance {R : Type*} [semiring R] [semiring k] [module R k] : module R (monoid_algebra k G) :=
+finsupp.module G k
```
I think this made more sense where it was, and reduces the chance of repeating work.

ocfnash replied:
Fair enough. I'll have to move the new stuff below the `derived_instances` section because it needs the `distrib_mul_action` instance (`module` was overkill). I'll try that and see how things look.

ocfnash committed:
Restore `monoid_algebra.module` to original location in `derived_instances` and move new content below.

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -250,6 +251,86 @@ finsupp.comap_distrib_mul_action_self
+begin
+  ext m,
+  classical,
+  simp only [mul_def, finsupp.mul_sum, smul_apply, sum_smul_index, implies_true_iff,
```
Are you sure you don't want to use `mul_apply` here, defined below? I think this stuff might still be too soon

ocfnash committed:
Move `non_unital_non_assoc_algebra` section.

Benefits:
 * Now adjacent to `algebra` section
 * We can use `monoid_algebra.mul_apply` (and so avoid `monoid_algebra.mul_def`)

ocfnash committed:
Suggestions of code review:

 * Move `monoid_algebra.of_magma`
 * Add `monoid_algebra.smul_comm_class_symm_self`

It's possible the tweak to use `simps` for the extant `monoid_algebra.of`
will break something but I'll have to wait and see.

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,81 @@ end
+
+/-- The functor `G ↦ monoid_algebra k G`, from the category of magmas to the category of non-unital,
+non-associative algebras over `k` is adjoint to the forgetful functor in the other direction. -/
+@[simps] def lift_magma {A : Type u₃}
```
If you generalize `lift_nc` to `non_unital_non_assoc_semiring`, can you use it here?

ocfnash replied:
Unfortunately the fact that `A` is non-unital makes it hard to factor `lift_magma` through `lift_nc` (more precisely, a generalisation of `lift_nc` that relaxes its `[semiring R]` argument to `[non_unital_non_assoc_semiring R]`).

The reason is that the defining property of `lift_nc` is:
```lean
@[simp] lemma lift_nc_single (f : k →+ R) (g : G →* R) (a : G) (b : k) :
  lift_nc f g (single a b) = f b * g a :=
...
```
whereas the defining property of `lift_magma` is:
```lean
variables {A : Type u₃} [non_unital_non_assoc_semiring A] [module k A] [is_scalar_tower k A A] [smul_comm_class k A A]
example (g : mul_hom G A) (a : G) (b : k) :
  lift_magma k g (single a b) = b • g a :=
...
```
I claim that to unify these we need a unit in `A` since we'd need to invoke `lift_nc` using `λ (b : k), (b • (1 : A))` for `f`.

I agree that with sufficient effort some unification between `lift_nc` and `lift_magma` could be obtained but the above makes me doubt that it is worth it.

eric-wieser replied:
That's a good point - I think we can still relax to remove associativity requirements from lift_nc, but I guess that's not in scope here.

eric-wieser replied:
Hmm, actually `lift_nc` seems to do exactly what you need here if you weaken its arguments to:
```lean
def lift_nc {R : Type*} [non_unital_non_assoc_semiring R] (f : k →+ R) (g : mul_hom G R) : monoid_algebra k G →+ R :=
lift_add_hom (λ x : G, (add_monoid_hom.mul_right (g x)).comp f)
```
You can then probably weakenn the subsequent lemmas to match.

ocfnash replied:
I'm sorry if I'm being dense here but I still don't see how to use `lift_nc` in `lift_magma`.

The problem is that when we are in the setting of `lift_magma`, because `A` has no unit, there is no way to construct the map `f` demanded by `lift_nc`.

I also note (though only as a secondary issue) that even without this problem, we could not use `lift_nc_mul` because it requires the multiplication in `R` to be associative (or at least to satisfy `mul_mul_mul_comm`).

I haven't looked through all the uses of `lift_nc` but I wouldn't be surprised if the right way resolve the overlap of responsibilities between `lift_magma` and `lift_nc` was to drop `lift_nc` entirely and make `lift_magma` the fundamental definition. In other words, the fundamental object into which we lift is a `non_unital_non_assoc_semiring R` but where `R` carries a compatible scalar action (as in `lift_magma`) rather than carrying a map `f : K → R` (as in `lift_nc`).

Please do enlighten me if I'm missing the obvious here.

eric-wieser replied:
I suspect it's me being dense after all - how about I make a PR with the generalizations I'm suggesting tomorrow, and then I'll think again about whether they're actually relevant here?

ocfnash replied:
By all means, if you have the appetite. I'll push on with some other work in an independent direction in the meantime. Thanks for all the great suggestions.

ocfnash committed:
Merge branch 'master' into magma_algebra2

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,81 @@ end
+
+variables (k) [semiring k] [has_mul G]
+
+instance is_scalar_tower_self : is_scalar_tower k (monoid_algebra k G) (monoid_algebra k G) :=
```
I think this might hold more generally for:
```suggestion
instance is_scalar_tower_self [is_scalar_tower R k k] : is_scalar_tower R (monoid_algebra k G) (monoid_algebra k G) :=
```
Although obviously that's not enough to make the build pass

eric-wieser replied:
Yeah, this works:
```lean
variables {R : Type*} (k) [semiring R] [semiring k] [module R k] [has_mul G]

instance is_scalar_tower_self [is_scalar_tower R k k] :
  is_scalar_tower R (monoid_algebra k G) (monoid_algebra k G) :=
⟨λ t a b, begin
  ext m,
  simp [mul_apply, finsupp.smul_sum, smul_ite, smul_mul_assoc, sum_smul_index'],
end⟩
```

eric-wieser replied:
This actually holds even for `distrib_mul_action R k` I think, but first `finsupp.smul_sum` needs its typeclasses relaxing (#8028)

ocfnash replied:
Done, and can confirm #8028 did suffice. Thanks!

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,81 @@ end
+
+instance smul_comm_class_symm_self {k : Type u₁} [comm_semiring k] :
+  smul_comm_class (monoid_algebra k G) k (monoid_algebra k G) :=
+smul_comm_class.symm k (monoid_algebra k G) (monoid_algebra k G)
```
These two fall out as something like:
```lean
instance smul_comm_class_self [smul_comm_class R k k] :
  smul_comm_class R (monoid_algebra k G) (monoid_algebra k G) :=
⟨λ t a b,
begin
  ext m,
  simp [mul_apply, finsupp.sum, finset.smul_sum, smul_ite, mul_smul_comm, sum_smul_index'],
end⟩

instance smul_comm_class_symm_self [smul_comm_class k R k] :
  smul_comm_class (monoid_algebra k G) R (monoid_algebra k G) :=
⟨λ t a b,
begin
  haveI : smul_comm_class R k k := smul_comm_class.symm _ _ _,
  rw ←smul_comm,
end⟩
```
I'm sure the second one can be golfed.

ocfnash replied:
Done. Couldn't see an obvious golfing but did reduce the character count.

ocfnash committed:
Merge branch 'master' into magma_algebra2

ocfnash committed:
Suggestions of CR: generalise instances.

Specifically the following:
 * `monoid_algebra.is_scalar_tower_self`
 * `monoid_algebra.smul_comm_class_self`
 * `monoid_algebra.smul_comm_class_symm_self`

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,85 @@ end
+          rw sum_add_index,
+          { intros m, exact zero_smul k (f m), },
+          { intros m t₁ t₂, exact add_smul t₁ t₂ (f m), },
+        end,
```
```suggestion
```

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,85 @@ end
+          have h₂ : ∀ m (t₁ t₂ : k), g m (t₁ + t₂) = g m t₁ + g m t₂, { intros, rw ← add_smul, },
+          simp_rw [finsupp.mul_sum, finsupp.sum_mul, smul_mul_smul, ← f.map_mul, mul_def,
+            sum_comm a₂ a₁, sum_sum_index h₁ h₂, sum_single_index (h₁ _)],
+        end, },
```
```suggestion
        end,
      .. lift_add_hom (λ x, (smul_add_hom k A).flip (f x)) },
```

ocfnash replied:
Nice

eric-wieser reviewed (commented):
A small golf:

ocfnash committed:
Update src/algebra/monoid_algebra.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Update src/algebra/monoid_algebra.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,79 @@ end
+      let F' := (F : monoid_algebra k G →+[k] A).to_add_monoid_hom,
+      change a.sum (λ m t, F' (finsupp.single m t)) = F' a,
+      rw [← add_monoid_hom.map_finsupp_sum F', sum_single],
+    end, }
```
I think this would become much shorter with a suitable `ext` lemma, which I think we should include as part of this PR:
```lean
variables {A : Type u₃}
  [non_unital_non_assoc_semiring A] [module k A] [is_scalar_tower k A A] [smul_comm_class k A A]

/-- A non_unital `k`-algebra homomorphism from `monoid_algebra k G` is uniquely defined by its
values on the functions `single a 1`. -/
lemma non_unital_alg_hom_ext ⦃φ₁ φ₂ : non_unital_alg_hom k (monoid_algebra k G) A⦄
  (h : ∀ x, φ₁ (single x 1) = φ₂ (single x 1)) : φ₁ = φ₂ :=
sorry
-- We should be able to adapt the proof of `alg_hom_ext` but many components of that are also missing


/-- See note [partially-applied ext lemmas]. -/
@[ext] lemma non_unital_alg_hom_ext' ⦃φ₁ φ₂ : non_unital_alg_hom k (monoid_algebra k G) A⦄
  (h : φ₁.to_mul_hom.comp (of_magma k G) = φ₂.to_mul_hom.comp (of_magma k G)) : φ₁ = φ₂ :=
non_unital_alg_hom_ext k $ mul_hom.congr_fun h
```

ocfnash replied:
Good point, I'll take this up later this morning.

ocfnash replied:
OK I've pushed https://github.com/leanprover-community/mathlib/pull/7932/commits/fbcff7e7de1db1521ad9ef0aef869ec63a740c36 to address this point.

I was a bit surprised that `distrib_mul_action_hom.to_linear_map` didn't already exist but it seems to be genuinely missing.

ocfnash committed:
Suggestion of CR: simplify `lift_magma` by using new ext lemma

ocfnash committed:
Fix long line

ocfnash committed:
Add lemma `distrib_mul_action_hom.coe_to_linear_map`

ocfnash committed:
Appease linter

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,85 @@ end
+non_unital_alg_hom.to_distrib_mul_action_hom_injective $
+distrib_mul_action_hom.to_linear_map_injective $
+finsupp.lhom_ext' $
+λ a, linear_map.ext_ring (h a)
```
I'd probably indent this to make it clear that wrapping is happening, either as:
```suggestion
non_unital_alg_hom.to_distrib_mul_action_hom_injective $
  distrib_mul_action_hom.to_linear_map_injective $
    finsupp.lhom_ext' $
      λ a, linear_map.ext_ring (h a)
```
or
```suggestion
non_unital_alg_hom.to_distrib_mul_action_hom_injective $
  distrib_mul_action_hom.to_linear_map_injective $
  finsupp.lhom_ext' $
  λ a, linear_map.ext_ring (h a)
```

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,85 @@ end
+/-- A non_unital `k`-algebra homomorphism from `monoid_algebra k G` is uniquely defined by its
+values on the functions `single a 1`. -/
+lemma non_unital_alg_hom_ext ⦃φ₁ φ₂ : non_unital_alg_hom k (monoid_algebra k G) A⦄
+  (h : ∀ x, φ₁ (single x 1) = φ₂ (single x 1)) : φ₁ = φ₂ :=
```
Does this generalize for `distrib_mul_action`? I haven't checked because gitpod is broken.

ocfnash replied:
Yes, I believe it is still true if we replace `[module k A]` with `[distrib_mul_action k A]`.

It looks like it might be quite a bit of work to generalise it since _I think_ almost all of the `finsupp` library demands a `module`, `linear_map` even when `distrib_mul_action`, `distrib_mul_action_hom` might suffice.

It's worth trying though. I'll have a go now.

ocfnash replied:
OK I have this working locally and it was not too much work. I'll tidy it up and push this evening.

ocfnash replied:
OK I just pushed https://github.com/leanprover-community/mathlib/pull/7932/commits/42ac4237f7ba07813b4e758bb231e1f0d6456843 which adds this generalisation.

ocfnash committed:
Generalise `non_unital_alg_hom_ext` from `module` to `distrib_mul_action`

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,86 @@ end
+/-- The functor `G ↦ monoid_algebra k G`, from the category of magmas to the category of non-unital,
+non-associative algebras over `k` is adjoint to the forgetful functor in the other direction. -/
+@[simps] def lift_magma [module k A] [is_scalar_tower k A A] [smul_comm_class k A A] :
+  mul_hom G A ≃ non_unital_alg_hom k (monoid_algebra k G) A :=
```
```lean
  mul_hom (multiplicative G) A ≃ non_unital_alg_hom k (add_monoid_algebra k G) A :=
```
Should be just as easy in the additive section, right?

ocfnash replied:
Sure.

ocfnash committed:
Add `add_monoid_algebra`-counterparts of `lift_magma` etc.

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -996,6 +1085,51 @@ namespace add_monoid_algebra
+direction. -/
+@[simps] def lift_magma [module k A] [is_scalar_tower k A A] [smul_comm_class k A A] :
+  mul_hom (multiplicative G) A ≃ non_unital_alg_hom k (add_monoid_algebra k G) A :=
+monoid_algebra.lift_magma k
```
I'm slightly worried that `simps` might generate nonsense here with badly handled `multiplicative`. What does this emit when you use `simps?`?

eric-wieser replied:
Yeah, this does generate nonsense:
```
#check lift_magma_apply_apply
```
```
lift_magma_apply_apply :
  ∀ (k : Type u_2) {G : Type u_3} [_inst_2 : semiring k] [_inst_4 : has_add G] {A : Type u_4}
  [_inst_5 : non_unital_non_assoc_semiring A] [_inst_6 : module k A] [_inst_7 : is_scalar_tower k A A]
  [_inst_8 : smul_comm_class k A A] (f : mul_hom (multiplicative G) A) (a : monoid_algebra k (multiplicative G)),
    (⇑(lift_magma k) f).to_fun a = sum a (λ (m : multiplicative G) (t : k), t • ⇑f m)
```
`a` has the wrong type

eric-wieser replied:
The lemmas you want there are:
```lean
lemma lift_magma_apply_apply [module k A] [is_scalar_tower k A A] [smul_comm_class k A A]
  (f : mul_hom (multiplicative G) A) (a : add_monoid_algebra k G) :
    lift_magma k f a = sum a (λ m t, t • f (multiplicative.of_add m)) := rfl

lemma lift_magma_symm_apply_apply [module k A] [is_scalar_tower k A A] [smul_comm_class k A A]
  (F : non_unital_alg_hom k (add_monoid_algebra k G) A) (a : multiplicative G) :
  (lift_magma k).symm F a = F (single a.to_add 1) := rfl
```

eric-wieser replied:
```suggestion
@[simps] def lift_magma [module k A] [is_scalar_tower k A A] [smul_comm_class k A A] :
  mul_hom (multiplicative G) A ≃ non_unital_alg_hom k (add_monoid_algebra k G) A :=
{ to_fun := λ f, { to_fun := λ a, sum a (λ m t, t • f (multiplicative.of_add m)),
                   .. (monoid_algebra.lift_magma k f : _)},
  inv_fun := λ F, F.to_mul_hom.comp (of_magma k G),
  .. (monoid_algebra.lift_magma k : mul_hom (multiplicative G) A ≃ non_unital_alg_hom k _ A) }
```
This generates the right simp lemmas, and also unfolds correctly too

ocfnash replied:
Thanks, I note that the final suggestion does not generate `lift_magma_symm_apply_apply` (as stated above) but that this is provable `by simp` via `lift_magma_symm_apply` (+ other basic lemmas) which it does generate, so all seems to be well.

eric-wieser reviewed (commented):
Do you think it's worth expanding the docstring a little to include some stuff about magma algebras?

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -373,6 +377,86 @@ end
+
+/-- The functor `G ↦ monoid_algebra k G`, from the category of magmas to the category of non-unital,
+non-associative algebras over `k` is adjoint to the forgetful functor in the other direction. -/
+@[simps] def lift_magma [module k A] [is_scalar_tower k A A] [smul_comm_class k A A] :
```
`simps` doesn't actually do the right thing on ~~`mul_hom`~~ `non_unital_alg_hom` just yet ~~- it's fine not to do it in this PR, but~~ (edit: this only applied to `mul_hom`, which was wrong) we're missing the configuration that tells `simp`s to use `coe_fn` instead of `.to_fun`.

ocfnash replied:
I'm having trouble parsing your words here but IIUC, `mul_hom` is not at fault, and `non_unital_alg_hom` is.

In any case, I can see the ugly `to_fun` in `lift_magma_apply_apply` but I can't see how to eliminate it. Based on [this comment](https://github.com/leanprover-community/mathlib/blob/1ffb5be5e1df05b9d571cc00b6a40b7d2f96c34e/src/algebra/group/hom.lean#L162) I wondered if the point might be to place `initialize_simps_projections non_unital_alg_hom (to_fun → apply)` lower down, but I saw no change when I moved the statement.

If this isn't easy to resolve, I'd be happy to drop `simps` here and just write the `simp` lemmas by hand.

eric-wieser replied:
Yes, `initialize_simps_projections non_unital_alg_hom (to_fun → apply)` needs to come _after_ the `has_coe_to_fun` instance.

ocfnash replied:
OK I'll move it now but just to repeat myself: this had no effect on `lift_magma_apply_apply`.

eric-wieser replied:
I found the cause I think - `non_unital_alg_hom` is defined for `monoid R`, but `has_coe_to_fun` is only defined for `semiring R`! ~~We should really fix that, but it doesn't really matter whether we do it in this PR or a pre-work / follow-up PR.~~ edit: It's a one line change, I just pushed it.

ocfnash replied:
Thanks so much for this!

ocfnash committed:
Update src/algebra/monoid_algebra.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

ocfnash committed:
Add doc-string-comment about the magma algebra

eric-wieser commented on src/algebra/monoid_algebra.lean:
```diff
@@ -15,6 +16,12 @@ a convolution product. To mathematicians this structure is known as the "monoid
+In fact the construction of the "monoid algebra" makes sense when `G` is not even a monoid, but
+merely a magma, i.e., when `G` carries a multiplication which is not required to satisfy any
+conditions at all. In this case the construction yields a not-necessarily-unital,
+not-necessarily-associative algebra but it is still adjoint to the forgetful functor from such
+algebras to magmas, and we prove this as `monoid_algebra.lift_magma`.
```
```suggestion
not-necessarily-associative algebra (one characterized by `[non_unital_non_assoc_semiring A]`
`[smul_comm_class R A A]` `[is_scalar_tower R A A]` instead of `[semiring A]` `[algebra R A]`)
but it is still adjoint to the forgetful functor from such algebras to magmas, and we prove this as
`monoid_algebra.lift_magma` (and the corresponding version for additive magmas
`add_monoid_algebra.lift_magma`).
```

ocfnash replied:
I'm not sure if this is helpful. I worry about drowning the reader in detail and after all, if they really need this, it's available in the type of `lift_magma`.

ocfnash replied:
And likewise, with the `add_monoid_algebra` point. It's certainly subjective and I'll happily accept the suggestion if you feel strongly, but I feel it's better to explain each point on its own and allow the consumer to assemble them to the particular situation they need.

eric-wieser replied:
Fair enough

ocfnash committed:
Move `initialize_simps_projections` for `non_unital_alg_hom`

eric-wieser committed:
fix(algebra/non_unital_alg_hom): relax typeclass in has_coe_to_fun

Previously `non_unital_alg_hom.has_coe_to_fun` only fired if `R` was a semiring, even though the object is permitted to exist when `R` is a monoid.

eric-wieser commented on src/algebra/non_unital_alg_hom.lean:
```diff
@@ -57,11 +57,9 @@ structure non_unital_alg_hom [monoid R]
 namespace non_unital_alg_hom
 
-variables {R A B C} [semiring R]
+variables {R A B C} [monoid R]
```
I think this will have made `simps` happy

ocfnash replied:
Boom! Nice find.

eric-wieser reviewed (approved):
bors r+

Thanks for perservering with this!
